### ISSUE TITLE
Add CPU hotplug support

### ISSUE THREAD

**Issue by ddingel** [docker/docker#27453]
### Description

Hotplugged CPUs are currently not recognized by Docker.

### Steps to reproduce the issue:

**Environment:**

The tests have been done on a s390x Ubuntu Xenial z/VM guest with three virtual CPUs and latest Docker built via `make deb`:

    root@r3545010:~# cat /etc/os-release 
    NAME="Ubuntu"
    VERSION="16.04.1 LTS (Xenial Xerus)"

    root@r3545010:~# uname -a
    Linux r3545010 4.4.0-21-generic #37-Ubuntu SMP Mon Apr 18 18:31:26 UTC 2016 s390x s390x s390x GNU/Linux

    root@r3545010:~# docker version
    Client:
     Version:      1.13.0-dev
     API version:  1.25
     Go version:   go1.7.1
     Git commit:   c505486
     Built:        Tue Sep 27 16:49:25 2016
     OS/Arch:      linux/s390x

**Test procedure:**

-   Stop docker daemon

-   Enable one CPU

-   Start docker daemon

-   Start first container and start two CPU burner processes

-   Enable second CPU

-   Check if second CPU is used by first container

-   Start second container and start two CPU burner processes

-   Check if second CPU is used by second container

**Describe the results you received:**

    root@r3545010:~# lscpu -e
    CPU NODE BOOK SOCKET CORE L1d:L1i:L2d:L2i ONLINE CONFIGURED POLARIZATION ADDRESS
    0   0    0    0      0    0:0:0:0         yes    yes        horizontal   0
    1   0    -    -      -    :::             no     yes        horizontal   1
    2   0    -    -      -    :::             no     yes        horizontal   2

    root@r3545010:~# systemctl start docker 

    root@r3545010:~# docker run -ti --rm rhel:7.1 bash

    bash-4.2# while [ true ]; do true; done &
    [1] 7
    bash-4.2# while [ true ]; do true; done &
    [2] 8

    root@r3545010:~# top
    top - 19:22:08 up 1 day,  1:13,  2 users,  load average: 0.59, 0.83, 0.78
    Tasks: 113 total,   3 running, 110 sleeping,   0 stopped,   0 zombie
    %Cpu0  : 97.4 us,  2.3 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.3 st
    KiB Mem :   881040 total,   385944 free,    78736 used,   416360 buff/cache
    KiB Swap:        0 total,        0 free,        0 used.   755592 avail Mem 

      PID USER      PR  NI    VIRT    RES    SHR S %CPU %MEM     TIME+ COMMAND
    23703 root      20   0    3536   1756   1416 R 50.0  0.2   0:08.83 bash
    23702 root      20   0    3536   1760   1420 R 49.7  0.2   0:10.43 bash

    root@r3545010:~# chcpu -e 1
    CPU 1 enabled

    root@r3545010:~# lscpu -e
    CPU NODE BOOK SOCKET CORE L1d:L1i:L2d:L2i ONLINE CONFIGURED POLARIZATION ADDRESS
    0   0    0    0      0    0:0:0:0         yes    yes        horizontal   0
    1   0    1    1      1    1:1:1:1         yes    yes        horizontal   1
    2   0    -    -      -    :::             no     yes        horizontal   2

    root@r3545010:~# top
    top - 19:23:19 up 1 day,  1:14,  2 users,  load average: 1.56, 1.08, 0.87
    Tasks: 110 total,   3 running, 107 sleeping,   0 stopped,   0 zombie
    %Cpu0  : 97.3 us,  2.3 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.3 st
    %Cpu1  :  0.0 us,  0.0 sy,  0.0 ni,100.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
    KiB Mem :   881040 total,   384772 free,    79844 used,   416424 buff/cache
    KiB Swap:        0 total,        0 free,        0 used.   754436 avail Mem 

      PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
    23703 root      20   0    3536   1756   1416 R  50.0  0.2   0:44.16 bash
    23702 root      20   0    3536   1760   1420 R  49.7  0.2   0:45.75 bash

Looking at the cpusets it turned out that docker and its containers only use CPU 0. The global cpuset also contains the new CPU 1:

    root@r3545010:/sys/fs/cgroup/cpuset# cat ./cpuset.cpus
    0-1
    root@r3545010:/sys/fs/cgroup/cpuset# cat ./docker/cpuset.cpus
    0
    root@r3545010:/sys/fs/cgroup/cpuset# cat ./docker/c22..217/cpuset.cpus
    0

It is possible to manually enable the CPU in the cpusets. First enable docker/cpuset.cpus and then do it for the container:

    root@r3545010:/sys/fs/cgroup/cpuset# echo "0-1" > ./docker/cpuset.cpus
    root@r3545010:/sys/fs/cgroup/cpuset# echo "0-1" > ./docker/c22..217/cpuset.cpus

After this change the container uses the new CPU:

    top - 19:55:28 up 1 day,  1:46,  2 users,  load average: 2.00, 1.99, 1.86
    Tasks: 110 total,   3 running, 107 sleeping,   0 stopped,   0 zombie
    %Cpu0  : 97.7 us,  2.3 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
    %Cpu1  : 97.7 us,  2.3 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
    KiB Mem :   881040 total,   384200 free,    80144 used,   416696 buff/cache
    KiB Swap:        0 total,        0 free,        0 used.   754060 avail Mem 

      PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
    23703 root      20   0    3536   1756   1416 R 100.0  0.2  19:46.40 bash
    23702 root      20   0    3536   1760   1420 R 100.0  0.2  19:47.84 bash

This can also be achived by using `docker update`, for example:

   `docker update --cpu-set-cpus 0-1 c22..217 `

When disabling and afterwards enabling CPU 1 again, the kernel (and Docker) forget the cpuset:

    root@r3545010:/sys/fs/cgroup/cpuset# chcpu -d 1
    CPU 1 disabled
    root@r3545010:/sys/fs/cgroup/cpuset# chcpu -e 1
    CPU 1 enabled
    root@r3545010:/sys/fs/cgroup/cpuset# cat docker/cpuset.cpus 
    0
    root@r3545010:/sys/fs/cgroup/cpuset# cat ./docker/c22..217/cpuset.cpus
    0

Also newly created containers only use CPU 0.

### Describe the results you expected:

Already running and newly started containers should be able to use additional online CPUs.

### Initial proposal
Containers which are not started with a `--cpuset-cpus` or `--cpuset-mems` option are called **unrestricted** in the following, the others are called **restricted**.

For each unrestricted container we skip the current creation of a cpuset cgroup sub-directory under the Docker cgroup directory. Then all tasks within that container will be automatically associated to the root cpuset cgroup. Because the root cpuset cgroup is managed by the kernel its cpuset.cpus will always contain all online CPUs. Therefore unrestricted containers will use all online CPUs.

This issue was created together with @michael-holzheu.

**Comment by michael-holzheu** [comment]
ping @justincormack

**Comment by justincormack** [comment]
This certainly seems a simple option. Is the code change in `runc`? If so might be worth opening an issue there?

**Comment by estesp** [comment]
Yeah, I think the actual change would be at the runc level, but seems appropriate for sure for Docker to handle  online CPU changes (hot plug).

**Comment by hqhq** [comment]
We had this discussing before, see if https://github.com/opencontainers/runc/pull/254 helps.

**Comment by ddingel** [comment]
So I created a new issue https://github.com/opencontainers/runc/issues/1119 on the runc project.

> We had this discussion before, see if opencontainers/runc#254 helps.

Thanks for the pointer. I read all comments as well as the code and think the conclusion at the time was right, and also that our proposal is different enough that it might be worth another discussion.

**Comment by kong62** [comment]
```
[root@Sit-Test-08 ~]# docker run -it harbor.sit.hupu.io/ci/jenkins:2.178 nproc                       
4

[root@Sit-Test-08 ~]# nproc 
8

```
[https://github.com/moby/moby/issues/40041](url)

**Comment by thaJeztah** [comment]
Did anyone try if this works with cgroupv2? https://github.com/opencontainers/runc/pull/254#issuecomment-138774034

/cc @AkihiroSuda
